File: Scripts/MelLog.cs
using UnityEngine;

namespace JJAudioUtils {
	public static class MelLog {

		/*
		 * Special thanks to Unity communities for solving MEL calculation issues!
		 */

		//returns the mel-adjusted float value given the samples, limits, and scale limit.
		public static float ExecuteMel(float[] samples, int lowerSampleLimit, int upperSampleLimit, float scaleLimit) {
			int _StrongestBin = GetStrongestBin (samples, lowerSampleLimit, upperSampleLimit );
			float _BinStrength = GetStrongestBinStrength (samples, lowerSampleLimit, upperSampleLimit );
			float val = Mathf.Clamp( ( ( _BinStrength * MelScale ( _BinStrength, _StrongestBin, samples.Length ) ) ), 0f, scaleLimit );
			return val;
		}
		
		public static float[] ExecuteLog(float[] samples) {
			for(int i = 0; i < samples.Length; i++) {
				if(i != 0)
					samples[i] *= (1 + Mathf.Log(i))/2;
				if(i == 0)
					samples[i] *= 1;
			}
			return samples;
		}


		//Logarithmically scales strength based on the frequency of the strongest sample to immitate human hearing range.
		//Return scaled strength.
		private static float MelScale ( float _Strength, int _StrongestBin, int sampleCounts ) {
			float _Frequency = _StrongestBin * ( 44100 / sampleCounts ) + 2000f / ( _StrongestBin + 1 );
			_Strength = 1127 * Mathf.Log ( 1 + ( _Frequency / 700f ) );

			return _Strength;
		}

		//Get strongest bin (the frequency range with the strongest signal) between _Bottom and _Top and return it.
		private static int GetStrongestBin ( float[] _Samples, int _Bottom, int _Top ) {
			float _Strength = 0f;
			int _StrongestIndex = 0;
			for ( int i = _Bottom; i < _Top; ++i ) {
				if ( _Strength < _Samples[ i ] ) {
					_Strength = _Samples[ i ];
					_StrongestIndex = i;
				}
			}
			return _StrongestIndex;
		}

		//Get strength of strongest bin and return it.
		private static float GetStrongestBinStrength ( float[] _Samples, int _Bottom, int _Top ) {
			float _Strength = 0f;
			for ( int i = _Bottom; i < _Top; ++i ) {
				if ( _Strength < _Samples[ i ] ) {
					_Strength = _Samples[ i ];
				}
			}
			return _Strength;
		}


	}
}

File: Scripts/MelLog.js
export const ExecuteMel = (samples_float_arr, lowerSampleLimit, upperSampleLimit, scaleLimit) => {
    let _strongestBin = GetStrongestBin(samples_float_arr, lowerSampleLimit, upperSampleLimit);
    let _binStrength = GetStrongestBinStrength(samples_float_arr, lowerSampleLimit, upperSampleLimit);
    let val = Math.min(Math.max(_binStrength * MelScaleStrength(_binStrength, _strongestBin, samples_float_arr.length), 0.0), scaleLimit);
    return val;
}

export const ExecuteLog = (samples_float_arr) => {
    for (let i = 0; i < samples_float_arr.length; i++) {
        if (i !== 0) {
            samples_float_arr[i] *= (1 + Math.log(i))/2;
        }
        if (i === 0) {
            samples_float_arr[i] *= 1;
        }
    }
    return samples_float_arr;
}

export const MelScaleStrength = (strength, strongest_bin, sample_counts) => {
    let _frequency = strongest_bin * (44100 / sample_counts) + 2000.0 / (strongest_bin + 1);
    strength = 1127 * Math.log(1 + (_frequency / 700.0));
    return strength;
}

export const GetStrongestBin = (samples_float_arr, bottom, top) => {
    let _strength = 0.0;
    let strongest_index = 0;
    for (let i = bottom; i < top; ++i) {
        if (_strength < samples_float_arr[i]) {
            _strength = samples_float_arr[i];
            strongest_index = i;
        }
    }
    return strongest_index;
}

export const GetStrongestBinStrength = (samples_float_arr, bottom, top) => {
    let _strength = 0.0;
    for (let i = bottom; i < top; ++i) {
        if (_strength < samples_float_arr[i]) {
            _strength = samples_float_arr[i];
        }
    }
    return _strength;
}
